onnxruntime/core/mlas/lib/q4_dq.cpp

```diff
@@ -294,3 +294,651 @@
+                            if (v < min) min = v;
+                            if (v > max) max = v;
+                        }
+                    }
+
+                    // store scale and zero point at quant parameter matrix position
+                    if (row_start < row_end) {
+                        const int32_t meta_idx = meta_col * row_blks + meta_row + kpack;
+                        if (zero_points == nullptr) {
+                            range2scale<ElementT, qbits>(min, max, scales[meta_idx]);
+                        } else {
+                            range2scalezp<ElementT, qbits>(min, max, scales[meta_idx], zp_bytes[kpack]);
+                        }
+                    }
+                }
+
+                // !! 4b specific code as we need to pack 2 4b numbers into one byte
+                if (zero_points != nullptr) {
+                    const int32_t meta_idx = meta_col * ((row_blks + 1) / 2) + meta_row / 2;
+                    zero_points[meta_idx] = (zp_bytes[0] & 0xf) | (zp_bytes[1] << 4);
+                }
+
+                for (int32_t j = c; j < c_end; ++j) {
+                    const int32_t meta_c = j / QuantBlk::kColumn;
+                    for (int32_t i = r; i < r_end; i += 2) {
+                        const int32_t meta_r = i / QuantBlk::kRow;
+                        const float scale = static_cast<float>(scales[meta_c * row_blks + meta_r]);
+                        const float reciprocal_scale = scale ? 1.0f / scale : 0.0f;
+                        const int8_t zp = zp_bytes[meta_r & 1];
+                        const int8_t zp1 = zp_bytes[((i + 1) / QuantBlk::kRow) & 1];
+
+                        const float v0 = static_cast<float>(src[i * leadingDimension + j]);
+                        const uint8_t vi0 =
+                            (uint8_t)std::min(BitsTraits<qbits>::kMaxFp,
+                                              std::max(0.0f, roundf(v0 * reciprocal_scale + zp)));
+
+                        uint8_t vi1 = (uint8_t)zp;
+                        if (i + 1 < r_end) {
+                            float reciprocal_scale1 = reciprocal_scale;
+                            if constexpr (QuantBlk::kRow == 1) {
+                                const float scale1 =
+                                    static_cast<float>(scales[meta_c * row_blks + meta_r + 1]);
+                                reciprocal_scale1 = scale1 ? 1.0f / scale1 : 0.0f;
+                            }
+                            const float v1 = static_cast<float>(src[(i + 1) * leadingDimension + j]);
+                            vi1 = (uint8_t)std::min(
+                                BitsTraits<qbits>::kMaxFp,
+                                std::max(0.0f, roundf(v1 * reciprocal_scale1 + zp1)));
+                        }
+
+                        // !! 4b specific code
+                        dst[j * q_rows + i / 2] = (vi0 & 0xf) | (vi1 << 4);
```

Comment:
>dst[j * q_rows + i / 2] = (vi0 & 0xf) | (vi1 << 4); [](http://example.com/codeflow?start=24&length=51)

How to store other bits, like 3, 5, 7?

Reply:
yeah the code is clearly marked as "4b specific" , we can discuss how to pack other bits later